### PR TITLE
chore: add postgres template caching for tests

### PR DESCRIPTION
**Pull request by hugodutka**
This PR is the first in a series aimed at closing [#15109](https://github.com/coder/coder/issues/15109).

### Changes

- **Template Database Creation:**  
  `dbtestutil.Open` now has the ability to create a template database if none is provided via `DB_FROM`. The template database’s name is derived from a hash of the migration files, ensuring that it can be reused across tests and is automatically updated whenever migrations change.

- **Optimized Database Handling:**  
  Previously, `dbtestutil.Open` would spin up a new container for each test when `DB_FROM` was unset. Now, it first checks for an active PostgreSQL instance on `localhost:5432`. If none is found, it creates a single container that remains available for subsequent tests, eliminating repeated container startups.

These changes address the long individual test times (10+ seconds) reported by some users, likely due to the time Docker took to start and complete migrations.

### CODE REVIEW

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:180]: Won't these panic if `cleanupConn` is nil? The `err` check above does not exit the function

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:177]: Sad we do not pass in `testing.T` to properly log this.

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:135]: Should this be `Open(t *testing.T, ...)` and not return a `cleanup` func?

Then we do `t.Cleanup(close())` inside this? I know it is more refactoring :cry:

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:243]: is 2137 just an arbitrary number? If so, can we drop a comment on it?

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:200]: Nice work thinking of the edge cases!

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:350]: tempDir is not used if the resource already exists

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:438]: If another process is reusing the same container, does this purge kill that second container?

Is the resource existence check for when you run `go test` in 2 different terminals?

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [inline coderd/database/dbtestutil/postgres.go:467]: Might be useful on the other docker create. If a test is `ctl+c` the cleanup will not run.

Although I do recall you mentioning the db **should** stay up after tests are done.

**Review by Emyrk — COMMENTED** [review]
Overall I like the direction for this.

The container staying up though I am more hesitant of. It just feels strange to leave open a container for running tests. But I also see the value in keeping it open :man_shrugging:.

I won't block on that. I am ok with keeping it, and letting people complain if it becomes an issue.

File: coderd/database/dbtestutil/postgres.go
Comment (hugodutka) [reply coderd/database/dbtestutil/postgres.go:83]: The container actually uses ramdisk so there's no volume. Cleaning it up would be also hard when you're running more than 1 test. You'd like to share it among the tests in that run, but clean up only once they all finish. And if you're running 2 `go test` processes at once, you'd need to make sure one doesn't kill the container while the other is running.

File: coderd/database/dbtestutil/postgres.go
Comment (hugodutka) [reply coderd/database/dbtestutil/postgres.go:438]: If another process is reusing the container there'd be only one container, and the purge would kill it.

> Is the resource existence check for when you run go test in 2 different terminals?

Yes.

File: coderd/database/dbtestutil/postgres.go
Comment (hugodutka) [reply coderd/database/dbtestutil/postgres.go:180]: good catch!

**Comment by hugodutka** [comment]
@Emyrk I addressed your comments. I don't think leaving a testing container open is an issue - that's how `make test-postgres` already works. Let's see how people react.

File: coderd/database/dbtestutil/postgres.go
Comment (Emyrk) [reply coderd/database/dbtestutil/postgres.go]: If test 1 calls purge while test 2 is running. That breaks test 2 right?

Comment (hugodutka) [reply coderd/database/dbtestutil/postgres.go]: That's correct. It would have to be a named container though, and the name would have to be the same in both runs. The test could work around this by generating a unique name per test run.

Comment (hugodutka) [reply coderd/database/dbtestutil/postgres.go]: Note: no tests currently use named containers because I introduced them in this PR :)

**Review by Emyrk — APPROVED** [review]
Approving. If people end up objecting about the docker container remaining, we can tweak.

Approving assuming CI does not break either :smile:

**Pull request merged**
